### PR TITLE
[onert] ShapeInference for Concat

### PR DESCRIPTION
**Pull request by hyunsik-yoon**
This adds shape inference for `Concat` op.

Signed-off-by: Hyun Sik Yoon <hyunsik.yoon.1024@gmail.com>

### CODE REVIEW

File: runtime/onert/core/src/util/shapeinf/Concat.cc [hunk 1]
```diff
@@ -0,0 +1,133 @@
...
+#include "util/ShapeInference.h"
+
+#include <cassert>
+
+namespace onert
+{
+namespace shape_inference
+{
+
+ir::Shape inferConcatShape(const Shapes &in_shapes, const ir::operation::Concat::Param &param)
```
Comment (hyunsik-yoon) [inline runtime/onert/core/src/util/shapeinf/Concat.cc:26]: moved from `ShapeInference.cc`

File: runtime/onert/core/src/util/shapeinf/Concat.cc [hunk 2]
```diff
@@ -0,0 +1,133 @@
...
+
+  // Calculate output shape
+  ir::Shape out_shape(first_in_shape);
+  out_shape.dim(concat_axis) = 0;
+  for (const auto &in_shape : in_shapes)
+    out_shape.dim(concat_axis) += in_shape.dim(concat_axis);
+  return out_shape;
+}
+
+void StaticInferer::visit(const ir::operation::Concat &op)
```
Comment (hyunsik-yoon) [inline runtime/onert/core/src/util/shapeinf/Concat.cc:47]: moved from `ShapeInference.cc`

**Comment by hyunsik-yoon** [comment]
People started to write shape inference of ops. I can review simple ops but it would take long time for me to review complex ops. :-) 
If @hseok-oh @glistening @dayo09 can review those, it would be wonderful since you know much more about complex ops. For that reason, I added you in the reviewer list. :-)

**Comment by glistening** [comment]
@hyunsik-yoon Perhaps I can start to read shape inference code soon. At last, optional tensor (fullyconnected) is almost done.

@wateret, @hseok-oh, @ragmani 

At a first glance on shape inference, I've found shape inference is under `util/shapeinf`.

Some passes are under `src/ir/pass` (e.g. `ConstantInsertionPass`, `PermutationOperationPass`.
Some are under `src/compiler` (e.g. `Fp32ToFp16Converter`, `OperationValidator`)
Shape inference is under `src/util/shapeinf`.

1. Why shape inference should be in `util`?
2. What is the difference between those under `pass` and `compiler`?

**Comment by hseok-oh** [comment]
>Why shape inference is under `src/util`

Last year, shape inference for some operation (for `inception` model) were implemented there.

**Comment by hyunsik-yoon** [comment]
@glistening 
Like @hseok-oh mentioned, last year `ShapeInference.h/cc` was in `util`.
I started to adding file into `ShapeInference.h/cc` and it was getting huge. 
We wanted to keep a file per each op not to make a huge file.

Basically one file per op comprises of the following:
1) shape calculation
2) shape inference during compilation phase (`StaticInferrer`)
3) shape inference during execution phase (`DynamicInferrer`)

Some design alternatives are:
a) since 1), 2), 3) are related keep them into one file
b) put 1) into `util`, put 2) into somewhere under `compiler`, and put 3) into somewhere under `exec`

If we prefer a),
I guess `util` is not a preferred location.
Since it lies on compilation and execution phase, we may want to create a new dir, e.g., `pass` (same level with `compilation` and `exec`) and the those files into `pass/shapeinf`.

Also some `*pass*`, and `*convert*`, which @glistening mentioned could be located under `pass`, if we consider them as _pass_.

(FYI, from the experience of compiler I saw that some passes has common logic at each phases. So compiler has pass as a separate directory. However we don't know clearly about onert since we don't have that many passes.)

pros: all related logic will be in a directory so easy to manage

if we prefer b),
we could split each inference file for each op into `util`, `compiler/pass`, and , `exec/pass`.

pros: better to understand which pass runs at each phase

I don't mind both ways but I now feel like b) seems better.

File: runtime/onert/core/src/util/shapeinf/Concat.cc [hunk 3]
```diff
@@ -0,0 +1,133 @@
...
+{
+  const int32_t concat_axis = param.axis;
+  const auto &first_in_shape = in_shapes[0];
+
+  // Check that all shapes are equal except for concat axis dimension
+  for (const auto &in_shape : in_shapes)
+  {
+    assert(in_shape.rank() == first_in_shape.rank());
+    for (int64_t dim_idx = 0; dim_idx < in_shape.rank(); ++dim_idx)
+      assert(dim_idx == concat_axis || in_shape.dim(dim_idx) == first_in_shape.dim(dim_idx));
```
Comment (ragmani) [inline runtime/onert/core/src/util/shapeinf/Concat.cc:36]: `concat_axis` may negative integer.

**Comment by ragmani** [comment]
> 2. What is the difference between those under `pass` and `compiler`?

@glistening 
I don't know about it. I just know that those under `pass` modify IR and update lowered information.

File: runtime/onert/core/src/util/shapeinf/Concat.cc (hunk 3 above)
Comment (hyunsik-yoon) [reply runtime/onert/core/src/util/shapeinf/Concat.cc]: Code considering axis < 0 and tests were added.

**Comment by hyunsik-yoon** [comment]
@glistening,
regarding https://github.com/Samsung/ONE/pull/1559#issuecomment-634995515,

How about discussing directory location issue as different issue?

**Comment by glistening** [comment]
@hyunsik-yoon Sure. I was just curious of the reason and history for this location. I don't insist to change the location in this PR.

**Comment by hyunsik-yoon** [comment]
PTAL

File: runtime/onert/core/src/util/shapeinf/Concat.cc [hunk 4]
```diff
@@ -0,0 +1,137 @@
...
+
+    in_shapes.emplace_back(shape);
+  }
+
+  auto output_shape = onert::shape_inference::inferConcatShape(in_shapes, op.param());
+
+  // set output shape and output buffer
+  setShape(output.get(), output_shape);
+
+  // assert(output->buffer() == nullptr);
```
Comment (ragmani) [inline runtime/onert/core/src/util/shapeinf/Concat.cc:132]: I found it

**Pull request merged**